### ISSUE TITLE
False positive needless_range_loop involving subslicing (`x[i..j]`)

### ISSUE THREAD

**Issue by kennytm** [rust-lang-nursery/rust-clippy#2213]
Test case:

```rust
fn main() {
    let mut g = vec![1, 2, 3, 4, 5, 6];
    let glen = g.len();
    for i in 0..glen {
        g[i] = g[i+1..].iter().sum();
    }
    assert_eq!(g, vec![20, 18, 15, 11, 6, 0]);
}
```

It produces this lint:

```
warning: the loop variable `i` is only used to index `g`.
 --> src/main.rs:4:5
  |
4 | /     for i in 0..glen {
5 | |         g[i] = g[i+1..].iter().sum();
6 | |     }
  | |_____^
  |
  = note: #[warn(needless_range_loop)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.169/index.html#needless_range_loop
help: consider using an iterator
  |
4 |     for <item> in g.iter().take(glen) {
  |         ^^^^^^
```

The suggestion is incorrect, it doesn't account for the `g[i + 1..]` part. Use `g.iter_mut().enumerate()` won't give a correct solution either, since you can't borrow `g` mutably and immutably simultaneously.

(Clippy version 0.0.169 as shown in the warning.)
